compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -57,117 +57,26 @@ object JavaNullInterop {
     assert(ctx.explicitNulls)
     assert(sym.is(JavaDefined), "can only nullify java-defined members")
 
-    // A list of "policies" that special-case certain members.
-    // The policies should be disjoint: we use the first one that is applicable.
-    val whitelist: Seq[NullifyPolicy] = Seq(
-      // The `TYPE` field in every class: don't nullify.
-      NoOpPolicy(_.name == nme.TYPE_),
-      // The `toString` method: don't nullify the return type.
-      paramsOnlyPolicy(_.name == nme.toString_),
-      // Constructors: params are nullified, but the result type isn't.
-      paramsOnlyPolicy(_.isConstructor),
-      // Java enum instances: don't nullify.
-      NoOpPolicy(_.isAllOf(Flags.JavaEnumValue)),
-      stdLibPolicy(sym, tp)
-    )
-    ++ (if (ctx.settings.YJavaInteropDontNullifyOutermost.value) {
-      Seq(
-        // Assume that all methods return non-nullable values.
-        paramsOnlyPolicy(_.is(Flags.Method)),
-        // Assume that all fields (static or not) contain non-nullable values.
-        // We assume that if a symbol is not a method then it is a field.
-        FieldPolicy(!_.is(Flags.Method))
-      )
-    } else {
-      Seq(FalsePolicy)
-    })
-
-    whitelist.find(_.isApplicable(sym)) match {
-      case Some(pol) => pol(tp)
-      case None => nullifyType(tp) // default case: nullify everything
-    }
+    // Some special cases when nullifying the type
+    if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
+      // Don't nullify the `TYPE` field in every class and Java enum instances
+      tp
+    else if (sym.name == nme.toString_ || sym.isConstructor || hasNotNullAnnot(sym))
+      // Don't nullify the return type of the `toString` method.
+      // Don't nullify the return type of constructors.
+      // Don't nullify the return type of methods with a not-null annotation.
+      nullifyExceptReturnType(tp)
+    else if (ctx.settings.YJavaInteropDontNullifyOutermost.value) 
+      nullifyExceptReturnType(tp)
```

Comment:
I think we can combine the method case and the field case now.

Reply:
Right, I agree.